tardis/io/atomic.py

```diff
@@ -482,113 +261,110 @@ def prepare_atom_data(self, selected_atomic_numbers, line_interaction_type='scat
-                    index=self.macro_atom_references.index))
-
-            self.macro_atom_data.insert(len(
-                self.macro_atom_data.columns), 'lines_idx',
-                pd.Series(self.lines_index.ix[self.macro_atom_data[
-                'transition_line_id']].values,
-                index=self.macro_atom_data.index))
-
-            tmp_lines_upper2level_idx = pd.MultiIndex.from_arrays(
-                [self.lines['atomic_number'], self.lines['ion_number'],
-                 self.lines['level_number_upper']])
-
-            self.lines_upper2macro_reference_idx = self.macro_atom_references['references_idx'].ix[
-                tmp_lines_upper2level_idx].values.astype(np.int64)
-
-            tmp_macro_destination_level_idx = pd.MultiIndex.from_arrays([self.macro_atom_data['atomic_number'],
-                                                                         self.macro_atom_data['ion_number'],
-                                                                         self.macro_atom_data[
-                                                                             'destination_level_number']])
+                self.macro_atom_references.loc[:, 'block_references'] = np.hstack(
+                    (0, np.cumsum(self.macro_atom_references['count_total'].values[:-1]))
+                )
+
+            self.macro_atom_references = self.macro_atom_references.set_index(
+                ['atomic_number', 'ion_number', 'source_level_number']
+            )
+
+            self.macro_atom_references.loc[:, "references_idx"] = np.arange(len(self.macro_atom_references))
+
+            self.macro_atom_data.loc[:, "lines_idx"] = self.lines_index.ix[
+                self.macro_atom_data['transition_line_id']
+            ].values
+
+            self.lines_upper2macro_reference_idx = self.macro_atom_references.ix[
+                tmp_lines_upper2level_idx, 'references_idx'
+            ].astype(np.int64).values
 
             if line_interaction_type == 'macroatom':
-                #Sets all
+                # Sets all
+                tmp_macro_destination_level_idx = pd.MultiIndex.from_arrays([
+                    self.macro_atom_data['atomic_number'],
+                    self.macro_atom_data['ion_number'],
+                    self.macro_atom_data['destination_level_number']
+                ])
 
-                self.macro_atom_data.insert(len(
-                    self.macro_atom_data.columns), 'destination_level_idx',
-                    pd.Series(self.macro_atom_references['references_idx'].ix[
-                    tmp_macro_destination_level_idx].values.astype(
-                        np.int64), index=self.macro_atom_data.index))
+                self.macro_atom_data.loc[:, 'destination_level_idx'] = self.macro_atom_references.ix[
+                    tmp_macro_destination_level_idx, "references_idx"
+                ].astype(np.int64).values  # why it is named `destination_level_idx` ?! It is reference index
```

Comment:
Hmm - not clear. Can you make sure to keep track of this question even if we merge